### DISCUSSION TITLE
File Uploads to outside container

### DISCUSSION THREAD

**Discussion by checkifexist** [ansibleguy76/ansibleforms#161] [Q&A] [answered]
Hi, Can we save file which user uploads from file form fields to a location outside container.
![image](https://github.com/ansibleguy76/ansibleforms/assets/157472645/0c95831c-6593-4556-9c9b-58160995068f)

i.e If a user upload a file, it should be uploaded to my local folder eg - ```/local/AF/files```
also this should be direct, so dont want to first upload to ```%PERSISTENT_FOLDER%/uploads``` then to ```/local/AF/files```

**Comment by ansibleguy76** [comment, marked as answer]
Use the UPLOAD_PATH and organize your environment (K8S, Docker, ...) so that it lands on the folder you want.
My choice of PERSISTENT_FOLDER is because that is generally a pass-through mapping in docker/k8s to an outside-container location (mother-host, nfs share, lun, ...)

The uploaded files will always have a temp name to avoid doubles incase multiple operator upload files.  The metadata is sent in the file-field (with temp en original name) so that you can trigger your own script/automation to rename and move the uploaded files.

In your case, it would be as simple as adding the following volumes mapping, in the assumption you use docker compose :

      - /local/AF/files:/app/dist/persistent/uploads

this will mount your local folder to the default upload location (you wouldn't even need to change the environment variable)

**Reply by checkifexist** [reply]
Got it thanks, and btw the names of uploaded files are random, so can we set it to the filename what user sends?

**Reply by ansibleguy76** [reply]
the upload is always under temp name (to avoid doubles) and the extravar for the file-field will contain the tmp name and the original name so that you can process it accordingly, rename it back to the original name, move the file, ...

**Reply by ansibleguy76** [reply]
If I would upload under the original name, the next issue would be that 2 operators upload a different file with the same name and the first upload is overwritten.  And the suggestion would then be : can't you use a tmp filename and send the original name so we can rename ourselves. =)

**Reply by checkifexist** [reply]
I understand, thankyou

**Reply by checkifexist** [reply]
@ansibleguy76 can we access files present in /persistent/uploads from other host. 
for eg- from a host i am trying to get this file ```/app/dist/persistent/uploads/file1.txt```, but the log says cannot access/find this file.

I tried to debug using localhost, it is not able to locate these folder. i.e when i tried ```ls /``` on localhost, there is no /app folder, so unable to find uploads. 

In short, How do i use uploaded files as src in ansible.builtin.copy

**Reply by ansibleguy76** [reply]
Hi,

So let's break this down again :

my host has the path `/srv/apps/ansibleforms-docker/data/uploads`
that path is mapped into the container as `/app/dist/persistent/uploads`

```
root@ubuntu:/srv/apps/ansibleforms-docker/data/uploads# ls
root@ubuntu:/srv/apps/ansibleforms-docker/data/uploads# touch test.txt
root@ubuntu:/srv/apps/ansibleforms-docker/data/uploads# ls
test.txt
root@ubuntu:/srv/apps/ansibleforms-docker/data/uploads#

```

let's go into the container and see if the test file is there :

```
root@ubuntu:/srv/apps/ansibleforms-docker/data/uploads# docker ps
CONTAINER ID   IMAGE                          COMMAND                  CREATED       STATUS      PORTS                                                  NAMES
6532973b2e2a   ansibleguy/ansibleforms:beta   "node ./dist/index.js"   4 weeks ago   Up 6 days   0.0.0.0:443->443/tcp, :::443->443/tcp                  ansibleforms-docker_app_1
4c3072080c09   mysql:8.1                      "docker-entrypoint.s…"   4 weeks ago   Up 6 days   0.0.0.0:3306->3306/tcp, :::3306->3306/tcp, 33060/tcp   ansibleforms-docker_mysqldb_1
root@ubuntu:/srv/apps/ansibleforms-docker/data/uploads# docker exec -it ansibleforms-docker_app_1 /bin/ash
/app # cd dist/persistent/uploads
/app/dist/persistent/uploads # ls
test.txt
/app/dist/persistent/uploads #
```

it is !

**Reply by ansibleguy76** [reply]
that was step 1, let's upload a file

```
name: Upload test
type: ansible
playbook: upload.yml
description: ""
roles:
  - public
categories: []
tileClass: has-background-info-light
icon: bullseye
fields:
  - name: file
    type: file
```

upload.yml :

```
---
- hosts: localhost
  tasks:
    - debug:
        var: file
    - name: make sure the target file does not exist
      ansible.builtin.file:
        path: /app/dist/persistent/uploads/{{ file.originalname }}
        state: absent
    - name: copy the file to originalname
      ansible.builtin.copy:
        src: "{{ file.path }}"
        dest: /app/dist/persistent/uploads/{{ file.originalname }}
    - name: remove the original file
      ansible.builtin.file:
        path: "{{ file.path }}"
        state: absent
```

**Reply by ansibleguy76** [reply]
```
[WARNING]: No inventory was parsed, only implicit localhost is available 2024-03-27 08:40:24
[WARNING]: provided hosts list is empty, only localhost is available. Note that 2024-03-27 08:40:24
the implicit localhost does not match 'all'

PLAY [localhost] *************************************************************** 2024-03-27 08:40:24

TASK [Gathering Facts] *********************************************************
ok: [localhost] 2024-03-27 08:40:25

TASK [debug] *******************************************************************
ok: [localhost] =>  2024-03-27 08:40:25
  file:
    destination: /app/dist/persistent/uploads
    encoding: 7bit
    fieldname: file
    filename: f61a1104d94e5078cacd0c02e64befec
    mimetype: text/html
    originalname: index.html
    path: /app/dist/persistent/uploads/f61a1104d94e5078cacd0c02e64befec
    size: 4485

TASK [make sure the target file does not exist] ********************************
ok: [localhost] 2024-03-27 08:40:25

TASK [copy the file to originalname] *******************************************
changed: [localhost] 2024-03-27 08:40:26

TASK [remove the original file] ************************************************
changed: [localhost] 2024-03-27 08:40:26

PLAY RECAP *********************************************************************
localhost                  : ok=5    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
ok: [Playbook finished] with status (0) 2024-03-27 08:40:26
```

**Reply by ansibleguy76** [reply]
In my container, the file is nicely there :
```

/app/dist/persistent/uploads # ls
index.html  test.txt
/app/dist/persistent/uploads #
```

let's go back to the host and look there:


```
root@ubuntu:/srv/apps/ansibleforms-docker/data/uploads# ls
index.html  test.txt
root@ubuntu:/srv/apps/ansibleforms-docker/data/uploads#
```


it's there.

let's see if we can access it from a remote machine:

```
$ ssh root@172.16.50.4
The authenticity of host '172.16.50.4 (172.16.50.4)' can't be established.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added '172.16.50.4' (ECDSA) to the list of known hosts.
root@172.16.50.4's password:
Welcome to Ubuntu 22.04.3 LTS (GNU/Linux 5.15.0-100-generic x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/pro

This system has been minimized by removing packages and content that are
not required on a system that users do not log into.

To restore this content, you can run the 'unminimize' command.
Last login: Wed Mar 27 08:36:52 2024 from 192.168.96.104
root@ubuntu:/home# cd /
root@ubuntu:/# cd /srv/apps/ansibleforms-docker/data/uploads
root@ubuntu:/srv/apps/ansibleforms-docker/data/uploads# ls
index.html  test.txt
root@ubuntu:/srv/apps/ansibleforms-docker/data/uploads#

```

there it is.

**Reply by ansibleguy76** [reply]
So summary :

/%your_host_path%/data/uploads => maps 1 on 1 with the /app/dist/persistent/uploads inside the container.  Every file written is seen in both locations.

you can remotely access your host by ssh and copy whatever file your want.

When you upload the file, you get all details and can rename/copy/move the file any way you want.

**Reply by checkifexist** [reply]
Thankyou for detail description of this. 
The files are available in both the location

but when i try to list the content of localhost using below plays 
![image](https://github.com/ansibleguy76/ansibleforms/assets/157472645/143cd1c5-6683-4cd8-93e0-99fe991d2c2e)

-> ```pwd``` shows ```/runner/project```
-> ```app/dist/persistent/uploads``` shows file/folder does not exist
-> ```ls /``` shows below files
![image](https://github.com/ansibleguy76/ansibleforms/assets/157472645/cc095f05-db37-4d56-872c-b8b2321591cc)

so although the uploaded files are in right place when i list from playbook it is not able to locate them, also when i went into ansible-form container using ```docker exec -it ansible-form sh```, files are correctly available in the location. its just that the localhost playbook are not able to access them

**Reply by ansibleguy76** [reply]
but you are sure you only use ansible within the container?  i have the feeling you have ansible also installed on the host os and try to access the container path from outside the container which obviously won't work. use the right path on the right side.  other wise print the os-version in the playbook. alpine is the containerversion.

**Reply by checkifexist** [reply]
I am using AWX, os version is showing linux

**Reply by checkifexist** [reply]
![image](https://github.com/ansibleguy76/ansibleforms/assets/157472645/9ea48641-3aeb-4d28-a715-1d9abdb0cbcf)

this is what i am mounting
